Refactored Plugin related code

Febbe opened the pull request:
- use std::unique_ptr
 - removed XournalType.h
 - fixed many clang-tidy warnings.
 - consistently use fs::path
 - mark functions override instead of virtual

rolandlo commented:
I was just looking over the code. Could you tell me, what you changed with your last commit/force-push? 
One remark: I needed to add `#include <memory>` to `PluginDialogEntry.h` to make it compile with gcc 9.2.1

Febbe commented:
> I was just looking over the code. Could you tell me, what you changed with your last commit/force-push?
> One remark: I needed to add `#include <memory>` to `PluginDialogEntry.h` to make it compile with gcc 9.2.1

I just added the #include <memory> in 2 files and fixed the clang-format. I normally use clang-12 to compile my cpp code. And clang is very friendly when it comes to missing header files. So it always compiles on my computer :D.

Febbe commented:
First of all I need feedback if all plugins still load and work as expected. I also need feedback if plugins in special non-ASCII folder names still work on Windows. Unfortunately I can't test this very often because my Windows PC is rarely accessible.

rolandlo commented:
> First of all I need feedback if all plugins still load and work as expected. I also need feedback if plugins in special non-ASCII folder names still work on Windows. Unfortunately I can't test this very often because my Windows PC is rarely accessible.

I have run your PR (compiled from source) from the terminal via `build/src/xournalpp`. All plugins in the `plugins` folder in the same `xournalpp` directory load as expected. To see filenames displayed lexically normal is certainly a plus. Otherwise all the plugins seem to work as before. 
One thing that bugs me is that accelerators like `<Release>z` don't seem to work. Firstly it is displayed like `Z` (as if there was no `<Release>` before the `z` key in the menu. Secondly pressing and releasing the `z` key does not execute the plugin. I don't know if I misunderstand something, but in the [GTK documentation](https://developer.gnome.org/gtk3/stable/gtk3-Keyboard-Accelerators.html#gtk-accelerator-parse) it is written 

> The format looks like “\<Control\>a” or “\<Shift\>\<Alt\>F1” or “\<Release\>z” (the last one is for key release).

So I would expect it to work.

Another question: Could we add a user plugin folder? Currently plugins are only loaded from one specific folder (which is a system folder in case Xournal++ has been installed on the system, and which cannot be modified without administrator rights).

I will now check how your PR works on Windows.

rolandlo commented:
Under windows there is something wrong with the terminal messages: 
![grafik](https://user-images.githubusercontent.com/40485433/99076716-dfdf5680-25bb-11eb-8c14-a6c66f53515f.png)

rolandlo commented:
Special folder names do not seem to work on Windows. I renamed the ColorCycle plugin to `àÄ@fix~d$$äö!è`, after which it was still correctly displayed in the Plugin Manager, but could not be loaded upon activating it and restarting Xournal++, see
![grafik](https://user-images.githubusercontent.com/40485433/99077216-ad822900-25bc-11eb-9909-1d4897cc7fe4.png)

Febbe commented:
@rolandlo the terminal in windows don't support utf8 by default. Sometimes it isn't even ASCII/ANSI. To fix that all << and >> musst be replaced by "

Febbe commented:
> One thing that bugs me is that accelerators like `<Release>z` don't seem to work. Firstly it is displayed like `Z` (as if there was no `<Release>` before the `z` key in the menu. Secondly pressing and releasing the `z` key does not execute the plugin. I don't know if I misunderstand something, but in the [GTK documentation](https://developer.gnome.org/gtk3/stable/gtk3-Keyboard-Accelerators.html#gtk-accelerator-parse) it is written

Has the issue been there before?  I only changed the path representation, the rest is code cleanup. I also don't quite understand what the problem is. What is the difference between accelerators like `<Release>z' and those that work.

rolandlo commented:
> Has the issue been there before? I only changed the path representation, the rest is code cleanup. I also don't quite understand what the problem is. What is the difference between accelerators like `z' and those that work.

Yes, this issue has been there before. The idea of using accelerators like `<Release>z` is for temporary activation of some tool. For instance you want to switch to the eraser tool when pressing `z` and switch back to the previously used tool when releasing this key. Accelerators work, but the key release accelerators (that should act upon releasing the key) don't.

Febbe commented:
> Another question: Could we add a user plugin folder? Currently plugins are only loaded from one specific folder (which is a system folder in case Xournal++ has been installed on the system, and which cannot be modified without administrator rights).

I think that's a great idea. There is already a user config folder in `~/.config/xournalpp` there should be another folder "plugins" in there, I can do this.

Febbe commented:
> Yes, this issue has been there before. The idea of using accelerators like `<Release>z` is for temporary activation of some tool. For instance you want to switch to the eraser tool when pressing `z` and switch back to the previously used tool when releasing this key. Accelerators work, but the key release accelerators (that should act upon releasing the key) don't.

I didn't wrote the plugin code, so I don't know where the accelerators are handled. I can imagine, that the accelerators won't be handled, if they are already handled. This must be investigated. You can add this to #2354 by editing my post.

Technius commented on src/plugin/Plugin.cpp:
```diff
@@ -282,33 +203,33 @@ void Plugin::loadScript() {
-        button.insert(std::pair<int, string>(0, _("OK")));
+    if (lua_pcall(lua.get(), 0, 0, 0) != LUA_OK) {
+        const char* errMsg = lua_tostring(lua.get(), -1);
+        map<int, std::string> button;
```
For consistency, this should also be `std::map`.

Technius commented on src/plugin/Plugin.cpp:
```diff
@@ -317,17 +238,14 @@ void Plugin::loadScript() {
-        button.insert(std::pair<int, string>(0, _("OK")));
+    if (lua_pcall(lua.get(), 0, 0, 0)) {
+        const char* errMsg = lua_tostring(lua.get(), -1);
+        map<int, std::string> button;
```
`std::map` here too.

Febbe replied:
done, also added a todo in our metaissue

Technius commented on src/plugin/Plugin.cpp:
```diff
@@ -16,66 +14,38 @@ extern "C" {
 
     if (lua_islightuserdata(lua, -1)) {
-        Plugin* data = (Plugin*)lua_touserdata(lua, -1);
+        auto* data = (Plugin*)lua_touserdata(lua, -1);
```
Or `static_cast`, whichever you think is more readable here (I have no opinion, they seem equally clear to me).

Technius commented:
> > Another question: Could we add a user plugin folder? Currently plugins are only loaded from one specific folder (which is a system folder in case Xournal++ has been installed on the system, and which cannot be modified without administrator rights).
> 
> I think that's a great idea. There is already a user config folder in `~/.config/xournalpp` there should be another folder "plugins" in there, I can do this.

See #1155 for the relevant issue.

Febbe commented:
> Special folder names do not seem to work on Windows. I renamed the ColorCycle plugin to `àÄ@fix~d$$äö!è`, after which it was still correctly displayed in the Plugin Manager, but could not be loaded upon activating it and restarting Xournal++, see
> ![grafik](https://user-images.githubusercontent.com/40485433/99077216-ad822900-25bc-11eb-9909-1d4897cc7fe4.png)
> 
> **Edit** even with a simple folder name like `René` or `Ähnlichkeit` the plugin cannot be loaded under Windows.

It seems like, this is a dependency issue caused by mingw-64. If the path exist, fopen should be able to open path::string::c_str
